File: lua/spacex-rest_sdk.lua
-- SpacexRest SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local SpacexRestSDK = {}
SpacexRestSDK.__index = SpacexRestSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

SpacexRestSDK._make_feature = _make_feature


function SpacexRestSDK.new(options)
  local self = setmetatable({}, SpacexRestSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features from config.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local feature_items = vs.items(feature_opts)
    if feature_items ~= nil then
      for _, item in ipairs(feature_items) do
        local fname = item[1]
        local fopts = helpers.to_map(item[2])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

  -- #BuildFeatures

  return self
end


function SpacexRestSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function SpacexRestSDK:get_utility()
  return Utility.copy(self._utility)
end


function SpacexRestSDK:get_root_ctx()
  return self._rootctx
end


function SpacexRestSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function SpacexRestSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:Capsule():list() / client:Capsule():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function SpacexRestSDK:Capsule(data)
  local EntityMod = require("entity.capsule_entity")
  if data == nil then
    if self._capsule == nil then
      self._capsule = EntityMod.new(self, nil)
    end
    return self._capsule
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Core():list() / client:Core():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function SpacexRestSDK:Core(data)
  local EntityMod = require("entity.core_entity")
  if data == nil then
    if self._core == nil then
      self._core = EntityMod.new(self, nil)
    end
    return self._core
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Crew():list() / client:Crew():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function SpacexRestSDK:Crew(data)
  local EntityMod = require("entity.crew_entity")
  if data == nil then
    if self._crew == nil then
      self._crew = EntityMod.new(self, nil)
    end
    return self._crew
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Landpad():list() / client:Landpad():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function SpacexRestSDK:Landpad(data)
  local EntityMod = require("entity.landpad_entity")
  if data == nil then
    if self._landpad == nil then
      self._landpad = EntityMod.new(self, nil)
    end
    return self._landpad
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Launch():list() / client:Launch():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function SpacexRestSDK:Launch(data)
  local EntityMod = require("entity.launch_entity")
  if data == nil then
    if self._launch == nil then
      self._launch = EntityMod.new(self, nil)
    end
    return self._launch
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Launchpad():list() / client:Launchpad():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function SpacexRestSDK:Launchpad(data)
  local EntityMod = require("entity.launchpad_entity")
  if data == nil then
    if self._launchpad == nil then
      self._launchpad = EntityMod.new(self, nil)
    end
    return self._launchpad
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Payload():list() / client:Payload():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function SpacexRestSDK:Payload(data)
  local EntityMod = require("entity.payload_entity")
  if data == nil then
    if self._payload == nil then
      self._payload = EntityMod.new(self, nil)
    end
    return self._payload
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Roadster():list() / client:Roadster():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function SpacexRestSDK:Roadster(data)
  local EntityMod = require("entity.roadster_entity")
  if data == nil then
    if self._roadster == nil then
      self._roadster = EntityMod.new(self, nil)
    end
    return self._roadster
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Rocket():list() / client:Rocket():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function SpacexRestSDK:Rocket(data)
  local EntityMod = require("entity.rocket_entity")
  if data == nil then
    if self._rocket == nil then
      self._rocket = EntityMod.new(self, nil)
    end
    return self._rocket
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Ship():list() / client:Ship():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function SpacexRestSDK:Ship(data)
  local EntityMod = require("entity.ship_entity")
  if data == nil then
    if self._ship == nil then
      self._ship = EntityMod.new(self, nil)
    end
    return self._ship
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Starlink():list() / client:Starlink():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function SpacexRestSDK:Starlink(data)
  local EntityMod = require("entity.starlink_entity")
  if data == nil then
    if self._starlink == nil then
      self._starlink = EntityMod.new(self, nil)
    end
    return self._starlink
  end
  return EntityMod.new(self, data)
end




function SpacexRestSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = SpacexRestSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return SpacexRestSDK

File: php/spacexrest_sdk.php
<?php
declare(strict_types=1);

// SpacexRest SDK

require_once __DIR__ . '/utility/struct/Struct.php';
require_once __DIR__ . '/core/UtilityType.php';
require_once __DIR__ . '/core/Spec.php';
require_once __DIR__ . '/core/Helpers.php';

// Load utility registration
require_once __DIR__ . '/utility/Register.php';

// Load config and features
require_once __DIR__ . '/config.php';
require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/features.php';

use Voxgig\Struct\Struct;

class SpacexRestSDK
{
    public string $mode;
    public array $features;
    public ?array $options;

    private $_utility;
    private $_rootctx;

    public function __construct(array $options = [])
    {
        $this->mode = "live";
        $this->features = [];
        $this->options = null;

        $utility = new SpacexRestUtility();
        $this->_utility = $utility;

        $config = SpacexRestConfig::make_config();

        $this->_rootctx = ($utility->make_context)([
            "client" => $this,
            "utility" => $utility,
            "config" => $config,
            "options" => $options ?? [],
            "shared" => [],
        ], null);

        $this->options = ($utility->make_options)($this->_rootctx);

        if (Struct::getpath($this->options, "feature.test.active") === true) {
            $this->mode = "test";
        }

        $this->_rootctx->options = $this->options;

        // Add features from config.
        $feature_opts = SpacexRestHelpers::to_map(Struct::getprop($this->options, "feature"));
        if ($feature_opts) {
            $items = Struct::items($feature_opts);
            if ($items) {
                foreach ($items as $item) {
                    $fname = $item[0];
                    $fopts = SpacexRestHelpers::to_map($item[1]);
                    if ($fopts && isset($fopts["active"]) && $fopts["active"] === true) {
                        ($utility->feature_add)($this->_rootctx, SpacexRestFeatures::make_feature($fname));
                    }
                }
            }
        }

        // Add extension features.
        $extend_val = Struct::getprop($this->options, "extend");
        if (is_array($extend_val)) {
            foreach ($extend_val as $f) {
                if (is_object($f) && method_exists($f, 'get_name')) {
                    ($utility->feature_add)($this->_rootctx, $f);
                }
            }
        }

        // Initialize features.
        foreach ($this->features as $f) {
            ($utility->feature_init)($this->_rootctx, $f);
        }

        ($utility->feature_hook)($this->_rootctx, "PostConstruct");
    }

    public function options_map(): array
    {
        $out = Struct::clone($this->options);
        return is_array($out) ? $out : [];
    }

    public function get_utility()
    {
        return SpacexRestUtility::copy($this->_utility);
    }

    public function get_root_ctx()
    {
        return $this->_rootctx;
    }

    public function prepare(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;
        $fetchargs = $fetchargs ?? [];

        $ctrl = SpacexRestHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "prepare",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $opts = $this->options;
        $path = Struct::getprop($fetchargs, "path") ?? "";
        $path = is_string($path) ? $path : "";
        $method_val = Struct::getprop($fetchargs, "method") ?? "GET";
        $method_val = is_string($method_val) ? $method_val : "GET";
        $params = SpacexRestHelpers::to_map(Struct::getprop($fetchargs, "params")) ?? [];
        $query = SpacexRestHelpers::to_map(Struct::getprop($fetchargs, "query")) ?? [];
        $headers = ($utility->prepare_headers)($ctx);

        $base = Struct::getprop($opts, "base") ?? "";
        $base = is_string($base) ? $base : "";
        $prefix = Struct::getprop($opts, "prefix") ?? "";
        $prefix = is_string($prefix) ? $prefix : "";
        $suffix = Struct::getprop($opts, "suffix") ?? "";
        $suffix = is_string($suffix) ? $suffix : "";

        $ctx->spec = new SpacexRestSpec([
            "base" => $base, "prefix" => $prefix, "suffix" => $suffix,
            "path" => $path, "method" => $method_val,
            "params" => $params, "query" => $query, "headers" => $headers,
            "body" => Struct::getprop($fetchargs, "body"),
            "step" => "start",
        ]);

        // Merge user-provided headers.
        $uh = Struct::getprop($fetchargs, "headers");
        if (is_array($uh)) {
            foreach ($uh as $k => $v) {
                $ctx->spec->headers[$k] = $v;
            }
        }

        [$_, $err] = ($utility->prepare_auth)($ctx);
        if ($err) {
            return ($utility->make_error)($ctx, $err);
        }

        [$fetchdef, $fd_err] = ($utility->make_fetch_def)($ctx);
        if ($fd_err) {
            return ($utility->make_error)($ctx, $fd_err);
        }
        return $fetchdef;
    }

    public function direct(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;

        // direct() is the raw-HTTP escape hatch: it never throws, it returns
        // an {ok, err, ...} dict. prepare() now raises on error, so catch it
        // and surface the failure through the dict instead.
        try {
            $fetchdef = $this->prepare($fetchargs);
        } catch (\Throwable $err) {
            return ["ok" => false, "err" => $err];
        }

        $fetchargs = $fetchargs ?? [];
        $ctrl = SpacexRestHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "direct",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $url = $fetchdef["url"] ?? "";
        [$fetched, $fetch_err] = ($utility->fetcher)($ctx, $url, $fetchdef);

        if ($fetch_err) {
            return ["ok" => false, "err" => $fetch_err];
        }

        if ($fetched === null) {
            return [
                "ok" => false,
                "err" => $ctx->make_error("direct_no_response", "response: undefined"),
            ];
        }

        if (is_array($fetched)) {
            $status = SpacexRestHelpers::to_int(Struct::getprop($fetched, "status"));
            $headers = Struct::getprop($fetched, "headers") ?? [];

            // No-body responses (204, 304) and explicit zero content-length
            // must skip JSON parsing — calling json() on an empty body errors.
            $content_length = is_array($headers) ? ($headers["content-length"] ?? null) : null;
            $no_body = $status === 204 || $status === 304 || (string)$content_length === "0";

            $json_data = null;
            if (!$no_body) {
                $jf = Struct::getprop($fetched, "json");
                if (is_callable($jf)) {
                    try {
                        $json_data = $jf();
                    } catch (\Throwable $e) {
                        // Non-JSON body — leave data null but keep status/ok.
                        $json_data = null;
                    }
                }
            }

            return [
                "ok" => $status >= 200 && $status < 300,
                "status" => $status,
                "headers" => Struct::getprop($fetched, "headers"),
                "data" => $json_data,
            ];
        }

        return [
            "ok" => false,
            "err" => $ctx->make_error("direct_invalid", "invalid response type"),
        ];
    }


    private $_capsule = null;

    // Canonical facade: $client->Capsule()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->capsule()
    // resolves here too.
    public function Capsule($data = null)
    {
        require_once __DIR__ . '/entity/capsule_entity.php';
        if ($data === null) {
            if ($this->_capsule === null) {
                $this->_capsule = new CapsuleEntity($this, null);
            }
            return $this->_capsule;
        }
        return new CapsuleEntity($this, $data);
    }


    private $_core = null;

    // Canonical facade: $client->Core()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->core()
    // resolves here too.
    public function Core($data = null)
    {
        require_once __DIR__ . '/entity/core_entity.php';
        if ($data === null) {
            if ($this->_core === null) {
                $this->_core = new CoreEntity($this, null);
            }
            return $this->_core;
        }
        return new CoreEntity($this, $data);
    }


    private $_crew = null;

    // Canonical facade: $client->Crew()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->crew()
    // resolves here too.
    public function Crew($data = null)
    {
        require_once __DIR__ . '/entity/crew_entity.php';
        if ($data === null) {
            if ($this->_crew === null) {
                $this->_crew = new CrewEntity($this, null);
            }
            return $this->_crew;
        }
        return new CrewEntity($this, $data);
    }


    private $_landpad = null;

    // Canonical facade: $client->Landpad()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->landpad()
    // resolves here too.
    public function Landpad($data = null)
    {
        require_once __DIR__ . '/entity/landpad_entity.php';
        if ($data === null) {
            if ($this->_landpad === null) {
                $this->_landpad = new LandpadEntity($this, null);
            }
            return $this->_landpad;
        }
        return new LandpadEntity($this, $data);
    }


    private $_launch = null;

    // Canonical facade: $client->Launch()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->launch()
    // resolves here too.
    public function Launch($data = null)
    {
        require_once __DIR__ . '/entity/launch_entity.php';
        if ($data === null) {
            if ($this->_launch === null) {
                $this->_launch = new LaunchEntity($this, null);
            }
            return $this->_launch;
        }
        return new LaunchEntity($this, $data);
    }


    private $_launchpad = null;

    // Canonical facade: $client->Launchpad()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->launchpad()
    // resolves here too.
    public function Launchpad($data = null)
    {
        require_once __DIR__ . '/entity/launchpad_entity.php';
        if ($data === null) {
            if ($this->_launchpad === null) {
                $this->_launchpad = new LaunchpadEntity($this, null);
            }
            return $this->_launchpad;
        }
        return new LaunchpadEntity($this, $data);
    }


    private $_payload = null;

    // Canonical facade: $client->Payload()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->payload()
    // resolves here too.
    public function Payload($data = null)
    {
        require_once __DIR__ . '/entity/payload_entity.php';
        if ($data === null) {
            if ($this->_payload === null) {
                $this->_payload = new PayloadEntity($this, null);
            }
            return $this->_payload;
        }
        return new PayloadEntity($this, $data);
    }


    private $_roadster = null;

    // Canonical facade: $client->Roadster()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->roadster()
    // resolves here too.
    public function Roadster($data = null)
    {
        require_once __DIR__ . '/entity/roadster_entity.php';
        if ($data === null) {
            if ($this->_roadster === null) {
                $this->_roadster = new RoadsterEntity($this, null);
            }
            return $this->_roadster;
        }
        return new RoadsterEntity($this, $data);
    }


    private $_rocket = null;

    // Canonical facade: $client->Rocket()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->rocket()
    // resolves here too.
    public function Rocket($data = null)
    {
        require_once __DIR__ . '/entity/rocket_entity.php';
        if ($data === null) {
            if ($this->_rocket === null) {
                $this->_rocket = new RocketEntity($this, null);
            }
            return $this->_rocket;
        }
        return new RocketEntity($this, $data);
    }


    private $_ship = null;

    // Canonical facade: $client->Ship()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->ship()
    // resolves here too.
    public function Ship($data = null)
    {
        require_once __DIR__ . '/entity/ship_entity.php';
        if ($data === null) {
            if ($this->_ship === null) {
                $this->_ship = new ShipEntity($this, null);
            }
            return $this->_ship;
        }
        return new ShipEntity($this, $data);
    }


    private $_starlink = null;

    // Canonical facade: $client->Starlink()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->starlink()
    // resolves here too.
    public function Starlink($data = null)
    {
        require_once __DIR__ . '/entity/starlink_entity.php';
        if ($data === null) {
            if ($this->_starlink === null) {
                $this->_starlink = new StarlinkEntity($this, null);
            }
            return $this->_starlink;
        }
        return new StarlinkEntity($this, $data);
    }



    public static function test(?array $testopts = null, ?array $sdkopts = null): self
    {
        $sdkopts = $sdkopts ?? [];
        $sdkopts = Struct::clone($sdkopts);
        $sdkopts = is_array($sdkopts) ? $sdkopts : [];

        $testopts = $testopts ?? [];
        $testopts = Struct::clone($testopts);
        $testopts = is_array($testopts) ? $testopts : [];
        $testopts["active"] = true;

        if (!isset($sdkopts["feature"])) {
            $sdkopts["feature"] = [];
        }
        $sdkopts["feature"]["test"] = $testopts;

        $sdk = new SpacexRestSDK($sdkopts);
        $sdk->mode = "test";
        return $sdk;
    }
}
